rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,227 @@
+changes and be able to handle them accordingly. Light clients should have the
+same version as the latest block protocol that they support. They will also
+require a similar method for verifying across block versions.
+
+#### Evidence
+
+Evidence is always a reference to a past action. It could be possible that the
+past action whether it be sending duplicate votes or attempting to fool a light
+client with an alternative block was of a different block version. To be able to
+process this correctly, evidence might need to include the version of it's
+contents or have a method of working out what version the contents should be in
+order to process them accordingly (upgrade heights could be stored in state).
+
+#### Spec
+
+In the future, we expect a versioned spec, where each spec release will list the
+software versions and protocol versions that correlate to that release. This
+will reside in `spec/version.md`. A clear guide of the rules behind breaking
+changes should also be there for easy reference.
+
+### Implementation Approach
+
+If this proposal is accepted, the implementation will need to be further fleshed
+out in an ADR, however, it is helpful to illustrate here the approach that
+Tendermint would most likely take to support soft upgrades.
+
+This would mostly likely mimic a similar design to what has been done in the
+Cosmos SDK with keeping legacy code and using conditionals to handle transitions
+between the versions. One would expect a directory structure for each module as
+such:
+
+```go
+consensus // module
+|---- state.go
+|---- state_test.go
+|       ...
+|---- votes.go // current version (v10)
+|       ...
+|---- legacy
+   |---- v8 // prior supported versions
+      |---- votes.go
+      |      ...    // other data structures ( and pure functions )
+   |---- v9
+      |---- votes.go
+      |      ...
+```
+
+The module, depending on the types, may statically cast each of the different
+versions or define an interface that it expects all versions to follow. For
+example:
+
+```go
+func (blockExec *BlockExecutor) CreateProposalBlock(state State) (BlockI, error)
+{
+ switch state.Version.Block {
+ case 9:
+  return v9.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ case 10:
+  return v10.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ default:
+  return fmt.Errorf("block version %d not recognised", state.Version.Block)
+ }
+}
+```
+
+In this case, Block which falls under a data structure of the block protocol
+could have many versions and is represented here as an interface. `State` and
+`BlockExecutor` are structs outside of the block protocol. Breaking changes to
+them could only occur in major releases.
+
+### Alternative Approach: Migration Scripts
+
+Although using migration scripts would have been most likely a simpler solution,
+unfortunately in most cases this would result in breaking the hashes and is
+therefore not a viable approach.
```

Comment:
Hm, I wonder if it might be possible to come up with something like a "derived" block, where we create a new version of a given height and somehow include the hash of the original block and the previous derived block. If the first block after the upgrade ties into this "derived" block rather than the original block, and we have the entire network come to consensus on it (and thus the entire history of derived blocks) we may get away with doing migrations, which I think are a much simpler solution to this problem.

Someone who knows more about this stuff than me should vet the idea, but maybe something worth exploring?

Reply:
So how would fast sync work in this case?

Reply:
It could fetch either the original or derived blocks, they both contain the transactions and related metadata, and are verifiable.

Reply:
This is an interesting idea!! One concern may be for the users that are storing blocks in a DB. They will need to rewrite the entire database. It could be considered a minor hiccup in the larger picture.

Reply:
So a node keeps both the original and derived block?

Reply:
> One concern may be for the users that are storing blocks in a DB. They will need to rewrite the entire database.

Yes, that's the tradeoff with migrations - they take time and disk space. But this is (mostly) a one-time cost, and I think the reduced complexity and operational burden is worth it.

> So a node keeps both the original and derived block?

That could be configurable, but some nodes in the network should keep the originals as well, for auditability. Keeping the original would also allow clients to request the block both in the original version and in the derived version, which would satisfy backwards compatibility with current clients.

Reply:
would this get compounded over multiple upgrades?

Reply:
>would this get compounded over multiple upgrades?

Depends on the design. The simplest solution would be to allow migrating from a derived block of the previous block version to a derived block of the current block version - this would only require us to support a n to n+1 version migration in the codebase, but nodes would have to rewrite the entire history for each upgrade.

Reply:
I'm just trying to wrap my brain around this. So in my head migration scripts wouldn't work because if for example we had a block at v1 and we wanted to migrate it to v2 where we decided to remove the evidence hash from the header, then when a new node was to fast sync, the signatures for the commit at that height would be aimed at a different block ID from the one that the new node would compute. How does this solve it?

Reply:
As I said, someone who knows more about this than me needs to vet this. But the derived block needs to contain some necessary data from the original block, such as the original block ID. And historical signatures will implicitly have been verified and signed by the current validators when they commit the block after the upgrade.

Reply:
and as well as containing the necessary data from the original block, should it also be able to form the new one?

I will mull this over a bit and see where it leads me. I guess my main concern is to remove the possibility that one node can fool a new node that is syncing up

Reply:
Have written this up in regards to doing migrations: https://github.com/tendermint/spec/pull/222#issuecomment-731139303